File: TypeScript/src/ArcscriptVisitor.ts
import BigNumber from 'bignumber.js';
import ArcscriptParserVisitor from './Generated/ArcscriptParserVisitor.js';
import ArcscriptFunctions, {
  ArcscriptNonVoidFunctionKeys,
  ArcscriptVoidFunctionKeys,
} from './ArcscriptFunctions.js';
import ArcscriptState from './ArcscriptState.js';
import { RuntimeError } from './errors/index.js';
import { MentionResult, VarObject, VarValue } from './types.js';
import {
  Additive_numeric_expressionContext,
  Argument_listContext,
  ArgumentContext,
  Assignment_segmentContext,
  BlockquoteContext,
  Compound_condition_andContext,
  Compound_condition_orContext,
  Conditional_sectionContext,
  ConditionContext,
  Else_if_clauseContext,
  Else_if_sectionContext,
  Else_sectionContext,
  ExpressionContext,
  Function_call_segmentContext,
  Function_callContext,
  If_clauseContext,
  If_sectionContext,
  InputContext,
  Mention_attributesContext,
  MentionContext,
  Multiplicative_numeric_expressionContext,
  Negated_unary_conditionContext,
  ParagraphContext,
  Script_sectionContext,
  Signed_unary_numeric_expressionContext,
  Statement_assignmentContext,
  Unary_conditionContext,
  Unary_numeric_expressionContext,
  Variable_listContext,
  Void_function_callContext,
} from './Generated/ArcscriptParser.js';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default class ArcscriptVisitor extends ArcscriptParserVisitor<any> {
  state: ArcscriptState;
  functions: ArcscriptFunctions;

  constructor(
    varValues: Record<string, VarValue>,
    varObjects: Record<string, VarObject>,
    elementVisits: Record<string, number>,
    currentElement: string,
    emit: (event: string, data: unknown) => void
  ) {
    super();

    this.state = new ArcscriptState(
      varValues,
      varObjects,
      elementVisits,
      currentElement,
      emit
    );

    this.functions = new ArcscriptFunctions(this.state);
  }

  visitInput = (ctx: InputContext) => {
    if (ctx.script()) {
      return { script: this.visitChildren(ctx.script()) };
    }

    return {
      condition: this.visitCompound_condition_or(ctx.compound_condition_or()),
    };
  };

  visitScript_section = (ctx: Script_sectionContext) => {
    if (!ctx) {
      return null;
    }

    if (
      Array.isArray(ctx.blockquote_list()) &&
      ctx.blockquote_list().length > 0
    ) {
      const result = ctx.blockquote_list().map(blockquoteContext => {
        return this.visitBlockquote(blockquoteContext);
      });
      return result;
    }

    if (
      Array.isArray(ctx.paragraph_list()) &&
      ctx.paragraph_list().length > 0
    ) {
      const result = ctx.paragraph_list().map(paragraphContext => {
        return this.visitParagraph(paragraphContext);
      });
      return result;
    }

    return this.visitChildren(ctx);
  };

  visitParagraph = (ctx: ParagraphContext) => {
    this.state.pushOutput(ctx.getText(), false);
    return ctx.getText();
  };

  visitBlockquote = (ctx: BlockquoteContext) => {
    this.state.addBlockquoteStart();
    this.visitChildren(ctx);
    this.state.addBlockquoteEnd();

    return ctx.getText();
  };

  visitAssignment_segment = (ctx: Assignment_segmentContext) => {
    this.state.addScript();
    return this.visitStatement_assignment(ctx.statement_assignment());
  };

  visitFunction_call_segment = (ctx: Function_call_segmentContext) => {
    this.state.addScript();
    return this.visitChildren(ctx.statement_function_call());
  };

  visitConditional_section = (ctx: Conditional_sectionContext) => {
    this.state.addScript();
    this.state.incrConditionDepth();
    const if_section = this.visitIf_section(ctx.if_section());
    if (if_section) {
      this.state.addScript();
      this.state.decrConditionDepth();
      return if_section;
    }
    const result = ctx.else_if_section_list().find(else_if_section => {
      const elif_section = this.visitElse_if_section(else_if_section);
      if (elif_section) {
        return true;
      }
      return false;
    });

    if (result) {
      this.state.addScript();
      this.state.decrConditionDepth();
      return result;
    }

    if (ctx.else_section()) {
      const elseResult = this.visitElse_section(ctx.else_section());
      this.state.addScript();
      this.state.decrConditionDepth();
      return elseResult;
    }
    this.state.addScript();
    this.state.decrConditionDepth();
    return null;
  };

  visitIf_section = (ctx: If_sectionContext) => {
    const result = this.visitIf_clause(ctx.if_clause());
    if (result) {
      return this.visitChildren(ctx.script());
    }

    return null;
  };

  visitElse_if_section = (ctx: Else_if_sectionContext) => {
    const result = this.visitElse_if_clause(ctx.else_if_clause());
    if (result) {
      return this.visitChildren(ctx.script());
    }
    return null;
  };

  visitElse_section = (ctx: Else_sectionContext) => {
    return this.visitChildren(ctx.script());
  };

  visitIf_clause = (ctx: If_clauseContext) => {
    return this.visitCompound_condition_or(ctx.compound_condition_or());
  };

  visitElse_if_clause = (ctx: Else_if_clauseContext) => {
    return this.visitCompound_condition_or(ctx.compound_condition_or());
  };

  visitStatement_assignment = (ctx: Statement_assignmentContext) => {
    const variableName = ctx.VARIABLE().getText();
    const variableObject = this.state.getVar(variableName);
    if (!variableObject) {
      throw new RuntimeError(`Variable ${variableName} not found`);
    }

    let variableValue = this.state.getVarValue(variableObject.id);
    let compound_condition_or = this.visitCompound_condition_or(
      ctx.compound_condition_or()
    );
    let result: VarValue = 0;
    if (
      typeof variableValue === 'string' ||
      typeof compound_condition_or === 'string'
    ) {
      if (ctx.ASSIGNADD()) {
        result = ((variableValue as string) + compound_condition_or) as string;
      } else if (ctx.ASSIGN()) {
        result = compound_condition_or;
      } else {
        throw new RuntimeError('Invalid operation with string');
      }
      this.state.setVarValues([variableObject.id], [result]);
      return null;
    }
    if (
      ctx.ASSIGNADD() ||
      ctx.ASSIGNSUB() ||
      ctx.ASSIGNMUL() ||
      ctx.ASSIGNMOD() ||
      ctx.ASSIGNDIV()
    ) {
      if (typeof variableValue === 'boolean') {
        variableValue = variableValue ? 1 : 0;
      }
      if (typeof compound_condition_or === 'boolean') {
        compound_condition_or = compound_condition_or ? 1 : 0;
      }
    }
    if (ctx.ASSIGNADD()) {
      result = new BigNumber(variableValue as number)
        .plus(new BigNumber(compound_condition_or as number))
        .toNumber();
    } else if (ctx.ASSIGNSUB()) {
      result = new BigNumber(variableValue as number)
        .minus(new BigNumber(compound_condition_or as number))
        .toNumber();
    } else if (ctx.ASSIGNMUL()) {
      result = new BigNumber(variableValue as number)
        .multipliedBy(new BigNumber(compound_condition_or as number))
        .toNumber();
    } else if (ctx.ASSIGNDIV()) {
      result = new BigNumber(variableValue as number)
        .dividedBy(new BigNumber(compound_condition_or as number))
        .toNumber();
      if (!Number.isFinite(result)) {
        throw new RuntimeError(`Invalid division by zero`);
      }
    } else if (ctx.ASSIGNMOD()) {
      result = new BigNumber(variableValue as number)
        .modulo(new BigNumber(compound_condition_or as number))
        .toNumber();
      if (!Number.isFinite(result)) {
        throw new RuntimeError(`Invalid division by zero`);
      }
    } else if (ctx.ASSIGN()) {
      result = compound_condition_or;
    }
    this.state.setVarValues([variableObject.id], [result]);
    return null;
  };

  visitVoid_function_call = (ctx: Void_function_callContext) => {
    let argument_list: (VarValue | MentionResult)[] = [];
    let fname: ArcscriptVoidFunctionKeys;
    if (ctx.VFNAME()) {
      fname = ctx.VFNAME().getText() as ArcscriptVoidFunctionKeys;
      if (ctx.argument_list()) {
        argument_list = this.visitArgument_list(ctx.argument_list());
      }
    }
    if (ctx.VFNAMEVARS()) {
      fname = ctx.VFNAMEVARS().getText() as ArcscriptVoidFunctionKeys;
      if (ctx.variable_list()) {
        argument_list = this.visitVariable_list(ctx.variable_list());
      }
    }

    this.functions[fname!](...argument_list);
  };

  visitVariable_list = (ctx: Variable_listContext): string[] => {
    return ctx.VARIABLE_list().map(variable => variable.getText());
  };

  visitCompound_condition_or = (
    ctx: Compound_condition_orContext
  ): VarValue => {
    const compound_condition_and = this.visitCompound_condition_and(
      ctx.compound_condition_and()
    );

    if (ctx.compound_condition_or()) {
      return (
        compound_condition_and ||
        this.visitCompound_condition_or(ctx.compound_condition_or())
      );
    }

    return compound_condition_and;
  };

  visitCompound_condition_and = (
    ctx: Compound_condition_andContext
  ): VarValue => {
    const negated_unary_condition = this.visitNegated_unary_condition(
      ctx.negated_unary_condition()
    );

    if (ctx.compound_condition_and()) {
      return (
        negated_unary_condition &&
        this.visitCompound_condition_and(ctx.compound_condition_and())
      );
    }

    return negated_unary_condition;
  };

  visitNegated_unary_condition = (ctx: Negated_unary_conditionContext) => {
    const unary_condition = this.visitUnary_condition(ctx.unary_condition());

    if (ctx.NEG() || ctx.NOTKEYWORD()) {
      return !unary_condition;
    }

    return unary_condition;
  };

  visitUnary_condition = (ctx: Unary_conditionContext) => {
    return this.visitCondition(ctx.condition());
  };

  visitCondition = (ctx: ConditionContext): VarValue => {
    if (ctx.expression_list().length === 1) {
      return this.visitExpression(ctx.expression(0));
    }
    const conditional_operator = ctx.conditional_operator();
    const exp0 = this.visitExpression(ctx.expression(0));
    const exp1 = this.visitExpression(ctx.expression(1));

    if (conditional_operator.GT()) {
      return exp0 > exp1;
    }
    if (conditional_operator.GE()) {
      return exp0 >= exp1;
    }
    if (conditional_operator.LT()) {
      return exp0 < exp1;
    }
    if (conditional_operator.LE()) {
      return exp0 <= exp1;
    }
    if (conditional_operator.EQ()) {
      return exp0 === exp1;
    }
    if (conditional_operator.NE()) {
      return exp0 !== exp1;
    }
    if (conditional_operator.ISKEYWORD()) {
      if (conditional_operator.NOTKEYWORD()) {
        return exp0 !== exp1;
      }
      return exp0 === exp1;
    }

    return this.visitChildren(ctx) as VarValue;
  };

  visitExpression = (ctx: ExpressionContext): VarValue => {
    if (ctx.STRING()) {
      let result = ctx.STRING().getText();
      result = result.replace(/^['"](.*)['"]$/, '$1');
      return result;
    }
    if (ctx.BOOLEAN()) {
      if (ctx.BOOLEAN().getText() === 'true') {
        return true;
      }
      return false;
    }

    return this.visitAdditive_numeric_expression(
      ctx.additive_numeric_expression()
    );
  };

  visitAdditive_numeric_expression = (
    ctx: Additive_numeric_expressionContext
  ): VarValue => {
    if (ctx.additive_numeric_expression()) {
      let additive_numeric_expression = this.visitAdditive_numeric_expression(
        ctx.additive_numeric_expression()
      );

      let multiplicative_numeric_expression =
        this.visitMultiplicative_numeric_expression(
          ctx.multiplicative_numeric_expression()
        );

      let hasString = typeof multiplicative_numeric_expression === 'string';

      hasString = hasString || typeof additive_numeric_expression === 'string';

      if (!hasString) {
        if (typeof multiplicative_numeric_expression === 'boolean') {
          multiplicative_numeric_expression = multiplicative_numeric_expression
            ? 1
            : 0;
        }
        if (typeof additive_numeric_expression === 'boolean') {
          additive_numeric_expression = additive_numeric_expression ? 1 : 0;
        }
      }

      if (ctx.ADD()) {
        if (hasString) {
          return ((additive_numeric_expression as string) +
            multiplicative_numeric_expression) as string;
        }
        return new BigNumber(additive_numeric_expression as number)
          .plus(new BigNumber(multiplicative_numeric_expression as number))
          .toNumber();
      }
      if (hasString) {
        throw new RuntimeError('Invalid subtraction with string');
      }
      return new BigNumber(additive_numeric_expression as number)
        .minus(new BigNumber(multiplicative_numeric_expression as number))
        .toNumber();
    }

    const multiplicative_numeric_expression =
      this.visitMultiplicative_numeric_expression(
        ctx.multiplicative_numeric_expression()
      );

    return multiplicative_numeric_expression;
  };

  visitMultiplicative_numeric_expression = (
    ctx: Multiplicative_numeric_expressionContext
  ): VarValue => {
    if (ctx.multiplicative_numeric_expression()) {
      let multiplicative_numeric_expression =
        this.visitMultiplicative_numeric_expression(
          ctx.multiplicative_numeric_expression()
        );
      let signed_unary_numeric_expression =
        this.visitSigned_unary_numeric_expression(
          ctx.signed_unary_numeric_expression()
        );
      if (typeof signed_unary_numeric_expression === 'boolean') {
        signed_unary_numeric_expression = signed_unary_numeric_expression
          ? 1
          : 0;
      }
      if (typeof multiplicative_numeric_expression === 'boolean') {
        multiplicative_numeric_expression = multiplicative_numeric_expression
          ? 1
          : 0;
      }
      if (ctx.MUL()) {
        return new BigNumber(multiplicative_numeric_expression)
          .multipliedBy(new BigNumber(signed_unary_numeric_expression))
          .toNumber();
      }
      if (ctx.DIV()) {
        const result = new BigNumber(multiplicative_numeric_expression)
          .dividedBy(new BigNumber(signed_unary_numeric_expression))
          .toNumber();
        if (!Number.isFinite(result)) {
          throw new RuntimeError(`Invalid division by zero`);
        }
        return result;
      }
      // else MOD
      const result = new BigNumber(multiplicative_numeric_expression)
        .modulo(new BigNumber(signed_unary_numeric_expression))
        .toNumber();
      if (!Number.isFinite(result)) {
        throw new RuntimeError(`Invalid division by zero`);
      }
      return result;
    }

    const signed_unary_numeric_expression =
      this.visitSigned_unary_numeric_expression(
        ctx.signed_unary_numeric_expression()
      );

    return signed_unary_numeric_expression;
  };

  visitSigned_unary_numeric_expression = (
    ctx: Signed_unary_numeric_expressionContext
  ): VarValue => {
    const unary_numeric_expression = this.visitUnary_numeric_expression(
      ctx.unary_numeric_expression()
    );

    const sign = ctx.sign();
    if (sign) {
      if (sign.ADD()) {
        return +unary_numeric_expression;
      }
      return -unary_numeric_expression;
    }
    return unary_numeric_expression;
  };

  visitUnary_numeric_expression = (
    ctx: Unary_numeric_expressionContext
  ): VarValue => {
    if (ctx.FLOAT()) {
      return parseFloat(ctx.FLOAT().getText());
    }
    if (ctx.INTEGER()) {
      return parseInt(ctx.INTEGER().getText(), 10);
    }
    if (ctx.VARIABLE()) {
      const variableName = ctx.VARIABLE().getText();
      const variableObject = this.state.getVar(variableName);
      if (!variableObject) {
        throw new RuntimeError(`Variable ${variableName} not found`);
      }
      return this.state.getVarValue(variableObject.id);
    }
    if (ctx.STRING()) {
      let result = ctx.STRING().getText();
      result = result.replace(/^['"](.*)['"]$/, '$1');
      return result;
    }
    if (ctx.BOOLEAN()) {
      if (ctx.BOOLEAN().getText() === 'true') {
        return true;
      }
      return false;
    }
    if (ctx.function_call()) {
      return this.visitFunction_call(ctx.function_call());
    }
    return this.visitCompound_condition_or(ctx.compound_condition_or());
  };

  visitFunction_call = (ctx: Function_callContext) => {
    let argument_list: (VarValue | MentionResult)[] = [];
    if (ctx.argument_list()) {
      argument_list = this.visitArgument_list(ctx.argument_list());
    }

    const function_name = ctx.FNAME().getText() as ArcscriptNonVoidFunctionKeys;
    return this.functions[function_name](...argument_list);
  };

  visitArgument_list = (ctx: Argument_listContext) => {
    return ctx.argument_list().map(argument => this.visitArgument(argument));
  };

  visitArgument = (ctx: ArgumentContext): VarValue | MentionResult => {
    if (ctx.STRING()) {
      let result = ctx.STRING().getText();
      result = result.replace(/^['"](.*)['"]$/, '$1');
      return result;
    }
    if (ctx.mention()) {
      return this.visitMention(ctx.mention());
    }
    return this.visitAdditive_numeric_expression(
      ctx.additive_numeric_expression()
    );
  };

  visitMention = (ctx: MentionContext): MentionResult => {
    const attrs: Record<string, string | boolean> = {};

    ctx.mention_attributes_list().forEach(attr => {
      const { name, value } = this.visitMention_attributes(attr);
      attrs[name] = value;
    });
    let label = '';
    if (ctx.MENTION_LABEL()) {
      label = ctx.MENTION_LABEL().getText();
    }

    return {
      attrs,
      label,
    };
  };

  visitMention_attributes = (
    ctx: Mention_attributesContext
  ): { name: string; value: boolean | string } => {
    const name = ctx.ATTR_NAME().getText();
    const ctxvalue = ctx.ATTR_VALUE();
    let value: boolean | string = true;
    if (ctxvalue) {
      value = ctxvalue.getText();
      if (value.startsWith('"') && value.endsWith('"')) {
        value = value.slice(1, -1);
      } else if (value.startsWith("'") && value.endsWith("'")) {
        value = value.slice(1, -1);
      }
    }

    return {
      name,
      value,
    };
  };
}

File: Cpp/src/ArcscriptVisitor.cpp
#include "antlr4-runtime.h"
#include "ArcscriptVisitor.h"
#include "ArcscriptParser.h"
#include "ArcscriptExpression.h"
#include "ArcscriptErrorExceptions.h"

using namespace Arcweave;

std::any ArcscriptVisitor::visitInput(ArcscriptParser::InputContext * ctx)
{
  if (ctx->script() != nullptr) {
    return visitScript(ctx->script());
  }
  // Condition
  auto comp_cond = std::any_cast<Expression>(visitCondition(ctx->condition()));
  return comp_cond.isTruthy();
}

std::any ArcscriptVisitor::visitCondition(ArcscriptParser::ConditionContext * ctx) {
  return visit(ctx->expression());
}

std::any ArcscriptVisitor::visitScript_section(ArcscriptParser::Script_sectionContext *ctx) {
  if (ctx == nullptr) {
    return {};
  }

  if (const auto blockquote_contexts = ctx->blockquote(); !blockquote_contexts.empty())
  {
    std::vector<std::any> result;
    for (const auto blockquote_context : blockquote_contexts)
    {
      result.push_back(visitBlockquote(blockquote_context));
    }
    return result;
  }

  if (const auto paragraph_contexts = ctx->paragraph(); !paragraph_contexts.empty())
  {
    std::vector<std::any> result;
    for (const auto paragraph_context : paragraph_contexts)
    {
      result.push_back(visitParagraph(paragraph_context));
    }
    return result;
  }

  return visitChildren(ctx);
}

std::any ArcscriptVisitor::visitBlockquote(ArcscriptParser::BlockquoteContext* context)
{
  state->outputs.AddBlockquote();
  visitChildren(context);
  state->outputs.ExitBlockquote();
  return context->getText();
}

std::any ArcscriptVisitor::visitParagraph(ArcscriptParser::ParagraphContext* context)
{
  const auto paragraph_end = context->PARAGRAPHEND()->getText();
  const auto paragraph_content = paragraph_end.substr(0, paragraph_end.size() - 4); // size of "</p>"
  state->outputs.AddParagraph(paragraph_content);
  return context->getText();
}


std::any ArcscriptVisitor::visitAssignment_segment(ArcscriptParser::Assignment_segmentContext *ctx) {
  state->outputs.AddScript();
  return visitStatement_assignment(ctx->statement_assignment());
}

std::any ArcscriptVisitor::visitFunction_call_segment(ArcscriptParser::Function_call_segmentContext *ctx) {
  state->outputs.AddScript();
  return visitFunction_call(ctx->function_call());
}

std::any ArcscriptVisitor::visitConditional_section(ArcscriptParser::Conditional_sectionContext *ctx) {
  state->outputs.AddScript();
  auto ifSection = std::any_cast<ConditionalSection>(visitIf_section(ctx->if_section()));
  if (ifSection.clause) {
    state->outputs.AddScript();
    return ifSection.script;
  }
  for (ArcscriptParser::Else_if_sectionContext *else_if_section : ctx->else_if_section()) {
    auto elif_section = std::any_cast<ConditionalSection>(visitElse_if_section(else_if_section));
    if (elif_section.clause) {
      return elif_section.script;
    }
  }
  if (ctx->else_section() != nullptr) {
    const auto elseSection = std::any_cast<ConditionalSection>(visitElse_section(ctx->else_section()));
    state->outputs.AddScript();
    return elseSection.script;
  }
  state->outputs.AddScript();
  return {};
}

std::any ArcscriptVisitor::visitIf_section(ArcscriptParser::If_sectionContext *ctx) {
  const auto result = std::any_cast<Expression>(visitIf_clause(ctx->if_clause()));
  ConditionalSection ifSection;
  ifSection.clause = false;
  if (result == true) {
    ifSection.clause = true;
    ifSection.script = visitScript(ctx->script());
  }
  return ifSection;
}

std::any ArcscriptVisitor::visitElse_if_section(ArcscriptParser::Else_if_sectionContext *ctx) {
  const auto result = std::any_cast<Expression>(visitElse_if_clause(ctx->else_if_clause()));
  ConditionalSection elseSection;
  elseSection.clause = false;
  if (result == true) {
    elseSection.clause = true;
    elseSection.script = visitScript(ctx->script());
  }

  return elseSection;
}

std::any ArcscriptVisitor::visitElse_section(ArcscriptParser::Else_sectionContext *ctx) {
  ConditionalSection elseIfSection;
  elseIfSection.clause = true;
  elseIfSection.script = visitScript(ctx->script());
  return elseIfSection;
}

std::any ArcscriptVisitor::visitIf_clause(ArcscriptParser::If_clauseContext *ctx) {
  return visit(ctx->expression());
}

std::any ArcscriptVisitor::visitElse_if_clause(ArcscriptParser::Else_if_clauseContext *ctx) {
  return visit(ctx->expression());
}

std::any ArcscriptVisitor::visitStatement_assignment(ArcscriptParser::Statement_assignmentContext *ctx) {
  auto identifier = std::any_cast<IdentifierDef>(visitAssignable(ctx->assignable()));
  auto identifierExpr = Expression();
  identifierExpr.setValue(state->getVarValue(identifier));

  auto re = std::any_cast<Expression>(visit(ctx->expression()));

  if (ctx->ASSIGN() != nullptr) {
    state->setVarValue(identifier, re.value);
    return {};
  }

  Expression varValue;
  varValue.setValue(state->getVarValue(identifier));

  if (ctx->ASSIGNADD() != nullptr) {
    varValue += re;
  } else if (ctx->ASSIGNSUB() != nullptr) {
    varValue -= re;
  } else if (ctx->ASSIGNMUL() != nullptr) {
    varValue *= re;
  } else if (ctx->ASSIGNDIV() != nullptr) {
    varValue /= re;
  } else if (ctx->ASSIGNMOD() != nullptr) {
    varValue %= re;
  }

  state->setVarValue(identifier, varValue.value);
  return {};
}

std::any ArcscriptVisitor::visitAssignable(ArcscriptParser::AssignableContext *ctx) {
  return visitIdentifier(ctx->identifier());
}

std::any ArcscriptVisitor::visitComparisonExpression(ArcscriptParser::ComparisonExpressionContext *ctx) {
  auto left = std::any_cast<Expression>(visit(ctx->expression(0)));
  if (ctx->AND() != nullptr || ctx->ANDKEYWORD() != nullptr) {
    if (!left.isTruthy()) {
      return Expression(false);
    }
    auto right = std::any_cast<Expression>(visit(ctx->expression(1)));
    return Expression(right.isTruthy());
  }

  if (ctx->OR() != nullptr || ctx->ORKEYWORD() != nullptr) {
    if (left.isTruthy()) {
      return Expression(true);
    }
    auto right = std::any_cast<Expression>(visit(ctx->expression(1)));
    return right;
  }

  auto right = std::any_cast<Expression>(visit(ctx->expression(1)));

  if (ctx->EQ() != nullptr || (ctx->ISKEYWORD() != nullptr && ctx->NOTKEYWORD() == nullptr)) {
    return Expression(left == right);
  }

  if (ctx->NE() != nullptr || (ctx->ISKEYWORD() != nullptr && ctx->NOTKEYWORD() != nullptr)) {
    return Expression(left != right);
  }

  if (ctx->LT() != nullptr) {
    return Expression(left < right);
  }

  if (ctx->GT() != nullptr) {
    return Expression(left > right);
  }

  if (ctx->LE() != nullptr) {
    return Expression(left <= right);
  }

  if (ctx->GE() != nullptr) {
    return Expression(left >= right);
  }

  throw RuntimeErrorException("Invalid comparison operator");
}

std::any ArcscriptVisitor::visitUnaryExpression(ArcscriptParser::UnaryExpressionContext *ctx) {
  if (ctx->NOTKEYWORD() != nullptr || ctx->NEG() != nullptr) {
    auto operand = std::any_cast<Expression>(visit(ctx->expression()));
    return Expression(!operand);
  }
  if (ctx->ADD() != nullptr) {
    return visit(ctx->expression());
  }
  if (ctx->SUB() != nullptr) {
    auto operand = std::any_cast<Expression>(visit(ctx->expression()));
    return -operand;
  }
  throw RuntimeErrorException("Invalid unary operator");
}

std::any ArcscriptVisitor::visitMultiplicativeExpression(ArcscriptParser::MultiplicativeExpressionContext *ctx) {
  auto left = std::any_cast<Expression>(visit(ctx->expression(0)));
  auto right = std::any_cast<Expression>(visit(ctx->expression(1)));
  if (ctx->MUL() != nullptr) {
    return left * right;
  }
  if (ctx->DIV() != nullptr) {
    return left / right;
  }
  if (ctx->MOD() != nullptr) {
    return left % right;
  }
  throw RuntimeErrorException("Invalid multiplicative operator");
}

std::any ArcscriptVisitor::visitAdditiveExpression(ArcscriptParser::AdditiveExpressionContext *ctx) {
  auto left = std::any_cast<Expression>(visit(ctx->expression(0)));
  auto right = std::any_cast<Expression>(visit(ctx->expression(1)));
  if (ctx->ADD() != nullptr) {
    return left + right;
  }
  if (ctx->SUB() != nullptr) {
    return left - right;
  }
  throw RuntimeErrorException("Invalid additive operator");
}

std::any ArcscriptVisitor::visitParenthesizedExpression(ArcscriptParser::ParenthesizedExpressionContext *ctx) {
  return visit(ctx->expression());
}

std::any ArcscriptVisitor::visitIdentifierExpression(ArcscriptParser::IdentifierExpressionContext *ctx) {
  auto idDef = std::any_cast<IdentifierDef>(visitIdentifier(ctx->identifier()));
  auto result = Expression();
  result.setValue(state->getVarValue(idDef));
  return result;
}

std::any ArcscriptVisitor::visitLiteralExpression(ArcscriptParser::LiteralExpressionContext *ctx) {
  return visitLiteral(ctx->literal());
}

std::any ArcscriptVisitor::visitFunctionCallExpression(ArcscriptParser::FunctionCallExpressionContext *ctx) {
  auto result = visitFunction_call(ctx->function_call());
  Expression e;
  e.setValue(result);
  return e;
}

std::any ArcscriptVisitor::visitLiteral(ArcscriptParser::LiteralContext *ctx) {
  if (ctx->STRING() != nullptr) {
    std::string result = ctx->STRING()->getText();
    result = result.substr(1, result.size() - 2);
    return Expression(result);
  }
  if (ctx->BOOLEAN() != nullptr) {
    return Expression(ctx->BOOLEAN()->getText() == "true");
  }
  return visitNumeric_literal(ctx->numeric_literal());
}

std::any ArcscriptVisitor::visitNumeric_literal(ArcscriptParser::Numeric_literalContext *ctx) {
  if (ctx->FLOAT() != nullptr) {
    return Expression(std::stod(ctx->FLOAT()->getText()));
  }
  return Expression(std::stoi(ctx->INTEGER()->getText()));;
}

std::any ArcscriptVisitor::visitFunction_call(ArcscriptParser::Function_callContext *ctx) {
  std::vector<std::any> argument_list_result;
  std::string fname = ctx->FNAME()->getText();

  if (ctx->argument_list() != nullptr) {
    argument_list_result = std::any_cast<std::vector<std::any>>(visitArgument_list(ctx->argument_list()));
  }

  if (ctx->identifier_list() != nullptr) {
    auto identifierList = std::any_cast<std::vector<IdentifierDef>>(visitIdentifier_list(ctx->identifier_list()));
    if (ArcscriptFunctions::functions[fname].argsType == "variable") {
      for (size_t i = 0; i < identifierList.size(); i++) {
        argument_list_result.push_back(state->getVar(identifierList[i]));
      }
    } else {
      for (size_t i = 0; i < identifierList.size(); i++) {
        auto e = Expression();
        e.setValue(state->getVarValue(identifierList[i]));
        argument_list_result.push_back(e);
      }
    }
  }

  std::any result = functions->Call(fname, argument_list_result);
  return result;
}

std::any ArcscriptVisitor::visitArgument_list(ArcscriptParser::Argument_listContext *ctx) {
  std::vector<std::any> arguments;
  for (ArcscriptParser::ArgumentContext *argument : ctx->argument()) {
    arguments.push_back(visitArgument(argument));
  }
  return arguments;
}

std::any ArcscriptVisitor::visitArgument(ArcscriptParser::ArgumentContext *ctx) {
  if (ctx->expression() != nullptr) {
    return visit(ctx->expression());
  }
  if (ctx->mention() != nullptr) {
    auto mention_result = std::any_cast<Mention>(visitMention(ctx->mention()));
    return mention_result;
  }

  throw RuntimeErrorException("Invalid function argument");
}

std::any ArcscriptVisitor::visitIdentifier_list(ArcscriptParser::Identifier_listContext *ctx) {
  std::vector<IdentifierDef> identifiers;
  for (auto *id : ctx->identifier()) {
    identifiers.push_back(std::any_cast<IdentifierDef>(visitIdentifier(id)));
  }
  return identifiers;
}

std::any ArcscriptVisitor::visitIdentifier(ArcscriptParser::IdentifierContext *ctx) {
  std::string name;
  std::string scope;
  if (ctx->IDENTIFIER().size() == 1) {
    name = ctx->IDENTIFIER(0)->getText();
  } else {
    scope = ctx->IDENTIFIER(0)->getText();
    name = ctx->IDENTIFIER(1)->getText();
  }
  IdentifierDef def;
  def.name = name;
  def.scope = scope;
  return def;
}

std::any ArcscriptVisitor::visitMention(ArcscriptParser::MentionContext *ctx) {
  std::map<std::string, std::string> attrs;
  for (ArcscriptParser::Mention_attributesContext *attr : ctx->mention_attributes()) {
    auto res = std::any_cast<std::map<std::string, std::string>>(visitMention_attributes(attr));
    attrs[res["name"]] = res["value"];
  }
  std::string label;
  if (ctx->MENTION_LABEL() != nullptr) {
    label = ctx->MENTION_LABEL()->getText();
  }
  Mention mention(label, attrs);
  return mention;
}

inline bool ends_with(std::string const & value, std::string const & ending)
{
    if (ending.size() > value.size()) return false;
    return std::equal(ending.rbegin(), ending.rend(), value.rbegin());
}

std::any ArcscriptVisitor::visitMention_attributes(ArcscriptParser::Mention_attributesContext *ctx) {
  std::string name = ctx->ATTR_NAME()->getText();
  antlr4::tree::TerminalNode *valueNode = ctx->ATTR_VALUE();
  std::string value(name);
  
  if (valueNode != nullptr) {
    std::string strvalue = valueNode->getText();
    if ((strvalue.rfind('\"', 0) == 0 && ends_with(strvalue, "\"")) ||
         (strvalue.rfind('\'', 0) == 0 && ends_with(strvalue, "'"))) {
      
      strvalue = strvalue.substr(1, strvalue.size() - 2);
      
    }
    value = strvalue;
  }
  return std::map<std::string, std::string> { {"name", name }, {"value", value } };
}
